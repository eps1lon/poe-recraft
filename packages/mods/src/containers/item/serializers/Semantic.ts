import Item from '../Item';
import { RarityIdent, RarityKind } from '../components/Rarity';
import { AugmentableValue } from '../util';
import { ModGroup } from './types';
import { BaseItemTypeProps, ModProps } from '../../../schema';
import { Mod } from '../../../mods';
import Container from '../../Container';
import { fromFlags } from '../atlasModifier';
import { Socket, SocketGroup } from '../components/Sockets';

/**
 * a serialization of an {Item} that aims convey meaning of its properties
 * and not how they are supposed to be displayed.
 *
 * Its goal is it to be deserialized into an Item
 */
export interface SemanticItemJSON {
  base: BaseItemTypeProps;
  name: string;
  rarity: RarityIdent;
  requirements: {
    level: AugmentableValue;
    dex: AugmentableValue;
    int: AugmentableValue;
    str: AugmentableValue;
  };
  mods?: { [key in ModGroup]?: ModProps[] };
  elder?: boolean;
  shaper?: boolean;
  corrupted?: boolean;
  mirrored?: boolean;
  item_level?: number;
  quality?: number;
  sockets: Socket[];
  groups: SocketGroup[];
}

export function serialize(item: Item): SemanticItemJSON {
  const base = item.baseitem;
  const [name, type_line] = item.name.lines();
  const rarity = item.rarity.toString();
  const requirements = item.requirements.list();
  const mods = {
    explicit: modProps(item.affixes),
    implicit: modProps(item.implicits),
  };
  const elder = item.isElderItem();
  const shaper = item.isSHaperItem();
  const { corrupted, item_level, mirrored } = item.props;
  const quality = item.properties.quality;
  const { sockets, groups } = serializeSockets(item);

  return {
    base,
    name: type_line !== undefined ? name : 'Random Name',
    rarity,
    requirements,
    mods,
    elder,
    shaper,
    corrupted,
    item_level,
    mirrored,
    quality,
    sockets,
    groups,
  };
}

export function deserialize(json: SemanticItemJSON): Item {
  const {
    mods = {},
    corrupted = false,
    mirrored = false,
    quality = 0,
    elder,
    shaper,
    sockets,
    groups,
    item_level = 100,
  } = json;
  const { explicit = [], implicit = [] } = mods;
  const atlas_modifier = fromFlags({ elder, shaper });

  const item = new Item({
    affixes: modsFromProps(explicit),
    baseitem: json.base,
    implicits: modsFromProps(implicit),
    meta_data: Item.buildMetaData(json.base),
    name: json.name,
    properties: { quality },
    props: { corrupted, mirrored, atlas_modifier, item_level },
    rarity: RarityKind[json.rarity],
    requirements: json.base.component_attribute_requirement,
    sockets: { sockets, groups },
  });
  return item;
}

function modProps(container: Container<Mod>): ModProps[] {
  return container.mods.map(({ props }) => props);
}

function modsFromProps(props: ModProps[]): Mod[] {
  return props.map(p => new Mod(p));
}

function serializeSockets(item: Item) {
  const json = item.sockets.toJson();
  const groups: number[][] = [];
  const sockets = json.map(({ group, ...socket }, index) => {
    if (!Array.isArray(groups[group])) {
      groups[group] = [];
    }
    groups[group].push(index);
    return socket;
  });

  return { sockets, groups };
}
