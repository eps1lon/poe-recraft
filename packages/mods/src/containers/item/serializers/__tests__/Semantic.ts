import { deserialize, serialize } from '../Semantic';
import item_fixtures from '../__fixtures__/items';

describe('Semantic', () => {
  for (const item of item_fixtures) {
    const snapshot_id = item.name.lines().join(' ');
    const serialized = serialize(item);
    test(`serialization matches snapshot ${snapshot_id}`, () => {
      expect(serialized).toMatchSnapshot();
    });
    test('deserialize() is the inverse of serialize()', () => {
      const deserialized = deserialize(serialized);
      expect(item.builder()).toMatchObject(deserialized.builder());
    });
  }
});
