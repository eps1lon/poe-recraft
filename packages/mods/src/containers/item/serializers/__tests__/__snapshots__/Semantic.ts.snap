// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Semantic serialization matches snapshot Journeyman's Platinum Kris 1`] = `
Object {
  "base": Object {
    "component_attribute_requirement": Object {
      "req_dex": 76,
      "req_int": 149,
      "req_str": 0,
    },
    "drop_level": 64,
    "height": 3,
    "id": "Metadata/Items/Weapons/OneHandWeapons/Daggers/Dagger20",
    "implicit_mods": Array [
      Object {
        "correct_group": "CriticalStrikeChanceIncrease",
        "domain": 1,
        "generation_type": 3,
        "id": "CriticalStrikeChanceImplicitDaggerNew3",
        "level": 1,
        "mod_type": Object {
          "primary": 185,
        },
        "name": "",
        "spawn_weights": Array [],
        "stat1_max": 50,
        "stat1_min": 50,
        "stat2_max": 0,
        "stat2_min": 0,
        "stat3_max": 0,
        "stat3_min": 0,
        "stat4_max": 0,
        "stat4_min": 0,
        "stat5_max": 0,
        "stat5_min": 0,
        "stats": Array [
          Object {
            "id": "critical_strike_chance_+%",
          },
        ],
        "tags": Array [],
      },
    ],
    "inherits_from": "Metadata/Items/Weapons/OneHandWeapons/Daggers/AbstractDagger",
    "item_class": "Dagger",
    "name": "Platinum Kris",
    "tags": Array [],
    "weapon_type": Object {
      "critical": 650,
      "damage_max": 95,
      "damage_min": 24,
      "range_max": 8,
      "speed": 833,
    },
    "width": 1,
  },
  "corrupted": false,
  "elder": false,
  "groups": Array [],
  "item_level": 100,
  "mirrored": true,
  "mods": Object {
    "explicit": Array [
      Object {
        "correct_group": "LocalIncreasedPhysicalDamagePercentAndAccuracyRating",
        "domain": 1,
        "generation_type": 1,
        "id": "LocalIncreasedPhysicalDamagePercentAndAccuracyRating2",
        "level": 11,
        "mod_type": Object {
          "primary": 11,
        },
        "name": "Journeyman's",
        "spawn_weights": Array [
          Object {
            "tag": "no_physical_damage_mods",
            "value": 0,
          },
          Object {
            "tag": "no_attack_mods",
            "value": 0,
          },
          Object {
            "tag": "weapon",
            "value": 1000,
          },
          Object {
            "tag": "default",
            "value": 0,
          },
        ],
        "stat1_max": 24,
        "stat1_min": 20,
        "stat2_max": 30,
        "stat2_min": 8,
        "stat3_max": 0,
        "stat3_min": 0,
        "stat4_max": 0,
        "stat4_min": 0,
        "stat5_max": 0,
        "stat5_min": 0,
        "stats": Array [
          Object {
            "id": "local_physical_damage_+%",
          },
          Object {
            "id": "local_accuracy_rating",
          },
        ],
        "tags": Array [
          "has_attack_mod",
        ],
      },
    ],
    "implicit": Array [
      Object {
        "correct_group": "CriticalStrikeChanceIncrease",
        "domain": 1,
        "generation_type": 3,
        "id": "CriticalStrikeChanceImplicitDaggerNew3",
        "level": 1,
        "mod_type": Object {
          "primary": 185,
        },
        "name": "",
        "spawn_weights": Array [],
        "stat1_max": 50,
        "stat1_min": 50,
        "stat2_max": 0,
        "stat2_min": 0,
        "stat3_max": 0,
        "stat3_min": 0,
        "stat4_max": 0,
        "stat4_min": 0,
        "stat5_max": 0,
        "stat5_min": 0,
        "stats": Array [
          Object {
            "id": "critical_strike_chance_+%",
          },
        ],
        "tags": Array [],
      },
    ],
  },
  "name": "Random Name",
  "quality": 0,
  "rarity": "magic",
  "requirements": Object {
    "dex": Value {
      "augmented": false,
      "classification": Array [],
      "modifiers": Array [],
      "range": ValueRange {
        "max": 76,
        "min": 76,
      },
    },
    "int": Value {
      "augmented": false,
      "classification": Array [],
      "modifiers": Array [],
      "range": ValueRange {
        "max": 149,
        "min": 149,
      },
    },
    "level": Value {
      "augmented": false,
      "classification": Array [],
      "modifiers": Array [],
      "range": ValueRange {
        "max": 64,
        "min": 64,
      },
    },
    "str": Value {
      "augmented": false,
      "classification": Array [],
      "modifiers": Array [],
      "range": ValueRange {
        "max": 0,
        "min": 0,
      },
    },
  },
  "shaper": false,
  "sockets": Array [],
}
`;

exports[`Semantic serialization matches snapshot Random Name Hubris Circlet 1`] = `
Object {
  "base": Object {
    "component_armour": Object {
      "armour": 0,
      "energy_shield": 76,
      "evasion": 0,
    },
    "component_attribute_requirement": Object {
      "req_dex": 0,
      "req_int": 154,
      "req_str": 0,
    },
    "drop_level": 69,
    "height": 2,
    "id": "Metadata/Items/Armours/Helmets/HelmetInt11",
    "implicit_mods": Array [],
    "inherits_from": "Metadata/Items/Armours/Helmets/AbstractHelmet",
    "item_class": "Helmet",
    "name": "Hubris Circlet",
    "tags": Array [
      "int_armour",
      "shaper_item",
      "helmet_shaper",
    ],
    "width": 2,
  },
  "corrupted": true,
  "elder": false,
  "groups": Array [],
  "item_level": 100,
  "mirrored": false,
  "mods": Object {
    "explicit": Array [
      Object {
        "correct_group": "IncreasedLife",
        "domain": 1,
        "generation_type": 1,
        "id": "IncreasedLife0",
        "level": 1,
        "mod_type": Object {
          "primary": 0,
        },
        "name": "Hale",
        "spawn_weights": Array [
          Object {
            "tag": "fishing_rod",
            "value": 0,
          },
          Object {
            "tag": "weapon",
            "value": 0,
          },
          Object {
            "tag": "default",
            "value": 1000,
          },
        ],
        "stat1_max": 9,
        "stat1_min": 3,
        "stat2_max": 0,
        "stat2_min": 0,
        "stat3_max": 0,
        "stat3_min": 0,
        "stat4_max": 0,
        "stat4_min": 0,
        "stat5_max": 0,
        "stat5_min": 0,
        "stats": Array [
          Object {
            "id": "base_maximum_life",
          },
        ],
        "tags": Array [],
      },
      Object {
        "correct_group": "BaseLocalDefences",
        "domain": 1,
        "generation_type": 1,
        "id": "LocalIncreasedEnergyShield2",
        "level": 11,
        "mod_type": Object {
          "primary": 7,
        },
        "name": "Glimmering",
        "spawn_weights": Array [
          Object {
            "tag": "int_armour",
            "value": 1000,
          },
          Object {
            "tag": "str_dex_int_armour",
            "value": 166,
          },
          Object {
            "tag": "default",
            "value": 0,
          },
        ],
        "stat1_max": 11,
        "stat1_min": 6,
        "stat2_max": 0,
        "stat2_min": 0,
        "stat3_max": 0,
        "stat3_min": 0,
        "stat4_max": 0,
        "stat4_min": 0,
        "stat5_max": 0,
        "stat5_min": 0,
        "stats": Array [
          Object {
            "id": "local_energy_shield",
          },
        ],
        "tags": Array [],
      },
      Object {
        "correct_group": "DefencesPercentAndStunRecovery",
        "domain": 1,
        "generation_type": 1,
        "id": "LocalIncreasedEnergyShieldPercentAndStunRecovery3",
        "level": 30,
        "mod_type": Object {
          "primary": 39,
        },
        "name": "Boggart's",
        "spawn_weights": Array [
          Object {
            "tag": "int_armour",
            "value": 1000,
          },
          Object {
            "tag": "default",
            "value": 0,
          },
        ],
        "stat1_max": 26,
        "stat1_min": 21,
        "stat2_max": 11,
        "stat2_min": 10,
        "stat3_max": 0,
        "stat3_min": 0,
        "stat4_max": 0,
        "stat4_min": 0,
        "stat5_max": 0,
        "stat5_min": 0,
        "stats": Array [
          Object {
            "id": "local_energy_shield_+%",
          },
          Object {
            "id": "base_stun_recovery_+%",
          },
        ],
        "tags": Array [],
      },
      Object {
        "correct_group": "FireResistance",
        "domain": 1,
        "generation_type": 2,
        "id": "FireResist1",
        "level": 1,
        "mod_type": Object {
          "primary": 96,
        },
        "name": "of the Whelpling",
        "spawn_weights": Array [
          Object {
            "tag": "weapon",
            "value": 500,
          },
          Object {
            "tag": "armour",
            "value": 1000,
          },
          Object {
            "tag": "ring",
            "value": 1000,
          },
          Object {
            "tag": "amulet",
            "value": 1000,
          },
          Object {
            "tag": "belt",
            "value": 1000,
          },
          Object {
            "tag": "quiver",
            "value": 1000,
          },
          Object {
            "tag": "default",
            "value": 0,
          },
        ],
        "stat1_max": 11,
        "stat1_min": 6,
        "stat2_max": 0,
        "stat2_min": 0,
        "stat3_max": 0,
        "stat3_min": 0,
        "stat4_max": 0,
        "stat4_min": 0,
        "stat5_max": 0,
        "stat5_min": 0,
        "stats": Array [
          Object {
            "id": "base_fire_damage_resistance_%",
          },
        ],
        "tags": Array [],
      },
      Object {
        "correct_group": "LocalAttributeRequirements",
        "domain": 1,
        "generation_type": 2,
        "id": "ReducedLocalAttributeRequirements1",
        "level": 36,
        "mod_type": Object {
          "primary": 898,
        },
        "name": "of the Worthy",
        "spawn_weights": Array [
          Object {
            "tag": "weapon",
            "value": 1000,
          },
          Object {
            "tag": "body_armour",
            "value": 850,
          },
          Object {
            "tag": "helmet",
            "value": 650,
          },
          Object {
            "tag": "shield",
            "value": 750,
          },
          Object {
            "tag": "gloves",
            "value": 450,
          },
          Object {
            "tag": "boots",
            "value": 550,
          },
          Object {
            "tag": "default",
            "value": 0,
          },
        ],
        "stat1_max": -18,
        "stat1_min": -18,
        "stat2_max": 0,
        "stat2_min": 0,
        "stat3_max": 0,
        "stat3_min": 0,
        "stat4_max": 0,
        "stat4_min": 0,
        "stat5_max": 0,
        "stat5_min": 0,
        "stats": Array [
          Object {
            "id": "local_attribute_requirements_+%",
          },
        ],
        "tags": Array [],
      },
      Object {
        "correct_group": "MineSupported",
        "domain": 1,
        "generation_type": 2,
        "id": "AdditionalMinesPlacedSupportedUber1_",
        "level": 68,
        "mod_type": Object {
          "primary": 4428,
        },
        "name": "of Shaping",
        "spawn_weights": Array [
          Object {
            "tag": "helmet_shaper",
            "value": 150,
          },
          Object {
            "tag": "default",
            "value": 0,
          },
        ],
        "stat1_max": 1,
        "stat1_min": 1,
        "stat2_max": 18,
        "stat2_min": 18,
        "stat3_max": 0,
        "stat3_min": 0,
        "stat4_max": 0,
        "stat4_min": 0,
        "stat5_max": 0,
        "stat5_min": 0,
        "stats": Array [
          Object {
            "id": "number_of_additional_mines_to_place",
          },
          Object {
            "id": "local_display_socketed_gems_get_remote_mine_level",
          },
        ],
        "tags": Array [],
      },
    ],
    "implicit": Array [],
  },
  "name": "Random Name",
  "quality": 0,
  "rarity": "rare",
  "requirements": Object {
    "dex": Value {
      "augmented": true,
      "classification": Array [],
      "modifiers": Array [],
      "range": ValueRange {
        "max": 0,
        "min": 0,
      },
    },
    "int": Value {
      "augmented": true,
      "classification": Array [],
      "modifiers": Array [],
      "range": ValueRange {
        "max": 126,
        "min": 126,
      },
    },
    "level": Value {
      "augmented": false,
      "classification": Array [],
      "modifiers": Array [],
      "range": ValueRange {
        "max": 69,
        "min": 69,
      },
    },
    "str": Value {
      "augmented": true,
      "classification": Array [],
      "modifiers": Array [],
      "range": ValueRange {
        "max": 0,
        "min": 0,
      },
    },
  },
  "shaper": true,
  "sockets": Array [],
}
`;
